src/storage/txn/txn_status_cache.rs

```diff
@@ -301,93 +382,182 @@ impl TxnStatusCache {
+                previous_size = large_txn_cache.size();
+                previous_allocated = large_txn_cache.internal_allocated_capacity();
+                if let Some(existing_entry) = large_txn_cache.get_mut(&start_ts) {
+                    // don't update committed or rolled back txns.
+                    if let TxnState::Ongoing {
+                        min_commit_ts: existing_min_commit_ts,
+                    } = existing_entry.state
+                    {
+                        if *min_commit_ts > existing_min_commit_ts {
+                            existing_entry.state = state;
+                            existing_entry.update_time = update_time;
+                        }
+                    }
+                } else {
+                    large_txn_cache.insert(start_ts, entry);
+                }
+                after_size = large_txn_cache.size();
+                after_allocated = large_txn_cache.internal_allocated_capacity();
+            }
+            // normal cache path
+            _ => {
+                let mut normal_cache = self.normal_cache[slot_index].lock();
+                previous_size = normal_cache.size();
+                previous_allocated = normal_cache.internal_allocated_capacity();
+                normal_cache.insert(start_ts, entry);
+                after_size = normal_cache.size();
+                after_allocated = normal_cache.internal_allocated_capacity();
+
+                let mut large_cache = self.large_txn_cache[slot_index].lock();
+                if let Some(existing_entry) = large_cache.get_mut(&start_ts) {
+                    // don't update committed or rolled back txns.
+                    if let TxnState::Ongoing { .. } = existing_entry.state {
+                        existing_entry.state = state;
+                        existing_entry.update_time = update_time;
+                    }
+                }
+            }
+        }
         // Update statistics.
         // CAUTION: Assuming that only one TxnStatusCache instance is in a TiKV process.
         SCHED_TXN_STATUS_CACHE_SIZE
             .used
-            .add(size as i64 - previous_size as i64);
+            .add(after_size as i64 - previous_size as i64);
         SCHED_TXN_STATUS_CACHE_SIZE
             .allocated
-            .add(allocated as i64 - previous_allocated as i64);
+            .add(after_allocated as i64 - previous_allocated as i64);
     }
 
-    /// Try to get an item from the cache, without promoting the item (if
-    /// exists) to the most recent place.
-    pub fn get_no_promote(&self, start_ts: TimeStamp) -> Option<TimeStamp> {
+    /// Insert a committed txn into the normal cache.
+    pub fn insert_committed(&self, start_ts: TimeStamp, commit_ts: TimeStamp, now: SystemTime) {
+        let state = TxnState::Committed { commit_ts };
+        self.upsert(start_ts, state, now);
+    }
+
+    pub fn get(&self, start_ts: TimeStamp) -> Option<TxnState> {
```

Comment:
The `pub` could be removed if it would not be used directly.

Reply:
This is an intended public interface of this struct, which will be used in resolved-ts resolver.